dbt-materialize: AssertionError: Expected count to be 1, but got 0 in test_full_deploy_process

### What version of Materialize are you using?

978d3323a464

### What is the issue?

Seen in https://buildkite.com/materialize/test/builds/84763#019059bb-1b80-479b-b7c6-1977ce3b9845
```
=================================== FAILURES ===================================
_______________ TestEndToEndDeployment.test_full_deploy_process ________________

self = <test_deploy.TestEndToEndDeployment object at 0xffff998b0d30>
project = <dbt.tests.fixtures.project.TestProjInfo object at 0xffff969a37c0>

    def test_full_deploy_process(self, project):
        # Prepare the source table, the sink cluster and schema
        project.run_sql("CREATE TABLE source_table (val INTEGER)")
        project.run_sql("CREATE CLUSTER sinks_cluster SIZE = '1'")
        project.run_sql("CREATE SCHEMA sinks_schema")

        project.run_sql("INSERT INTO source_table VALUES (1)")
        run_dbt(["run"])

        created_schema = project.created_schemas[0]

        project_config = f"{{deployment: {{default: {{clusters: ['quickstart'], schemas: ['{created_schema}']}}}}}}"
        project_config_deploy = f"{{deployment: {{default: {{clusters: ['quickstart'], schemas: ['{created_schema}']}}}}, deploy: True}}"

        # Validate the initial sink result
        project.run_sql(
            "CREATE SOURCE sink_validation_source FROM KAFKA CONNECTION kafka_connection (TOPIC 'testdrive-test-sink-1') FORMAT JSON"
        )
        time.sleep(3)
        result = project.run_sql(
            "SELECT count(*) FROM sink_validation_source", fetch="one"
        )
>       assert result[0] == 1, f"Expected count to be 1, but got {result[0]}"
E       AssertionError: Expected count to be 1, but got 0
E       assert 0 == 1

/dbt-materialize/tests/adapter/test_deploy.py:643: AssertionError
```
This test was added in https://github.com/MaterializeInc/materialize/pull/27557

Maybe we require some retry logic here, similar to what we have in Testdrive?

--- a/misc/dbt-materialize/tests/adapter/test_deploy.py
+++ b/misc/dbt-materialize/tests/adapter/test_deploy.py
@@ -636,19 +636,15 @@ class TestEndToEndDeployment:
         project.run_sql(
             "CREATE SOURCE sink_validation_source FROM KAFKA CONNECTION kafka_connection (TOPIC 'testdrive-test-sink-1') FORMAT JSON"
         )
-        time.sleep(3)
-        result = project.run_sql(
-            "SELECT count(*) FROM sink_validation_source", fetch="one"
+        run_with_retry(
+            project, "SELECT count(*) FROM sink_validation_source", expected_count=1
         )
-        assert result[0] == 1, f"Expected count to be 1, but got {result[0]}"
 
         # Insert another row and validate the sink
         project.run_sql("INSERT INTO source_table VALUES (2)")
-        time.sleep(3)
-        result = project.run_sql(
-            "SELECT count(*) FROM sink_validation_source", fetch="one"
+        run_with_retry(
+            project, "SELECT count(*) FROM sink_validation_source", expected_count=2
         )
-        assert result[0] == 2, f"Expected count to be 2, but got {result[0]}"
 
         # Initialize the deployment environment
         run_dbt(["run-operation", "deploy_init", "--vars", project_config])
@@ -665,19 +661,15 @@ class TestEndToEndDeployment:
         )
 
         # Ensure the validation source has not changed
-        result = project.run_sql(
-            "SELECT count(*) FROM sink_validation_source", fetch="one"
+        run_with_retry(
+            project, "SELECT count(*) FROM sink_validation_source", expected_count=2
         )
-        time.sleep(3)
-        assert result[0] == 2, f"Expected count to be 2, but got {result[0]}"
 
         # Insert a new row and validate the new sink result after the deploy
         project.run_sql("INSERT INTO source_table VALUES (3)")
-        time.sleep(3)
-        result = project.run_sql(
-            "SELECT count(*) FROM sink_validation_source", fetch="one"
+        run_with_retry(
+            project, "SELECT count(*) FROM sink_validation_source", expected_count=3
         )
-        assert result[0] == 3, f"Expected count to be 3, but got {result[0]}"
 
         # Get the IDs of the materialized views
         before_view_id = project.run_sql(
@@ -742,3 +734,14 @@ class TestEndToEndDeployment:
         ), "Sink's view ID should be different after deployment"
 
         run_dbt(["run-operation", "deploy_cleanup", "--vars", project_config])
+
+
+def run_with_retry(project, sql, expected_count, retries=5, delay=3, fetch="one"):
+    for i in range(retries):
+        result = project.run_sql(sql, fetch=fetch)
+        if result[0] == expected_count:
+            return result
+        time.sleep(delay)
+    raise AssertionError(
+        f"Expected count to be {expected_count}, but got {result[0]} after {retries} retries"
+    )

